frontend/lib/src/components/elements/ArrowVegaLiteChart/useVegaEmbed.ts

```diff
@@ -0,0 +1,285 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { RefObject, useCallback, useEffect, useRef } from "react"
+
+import { truthy, View as VegaView } from "vega"
+import embed from "vega-embed"
+import { expressionInterpreter } from "vega-interpreter"
+
+import { useFormClearHelper } from "@streamlit/lib/src/components/widgets/Form"
+import { WidgetStateManager } from "@streamlit/lib/src/WidgetStateManager"
+import { Quiver } from "@streamlit/lib/src/dataframes/Quiver"
+import { logMessage, logWarning } from "@streamlit/lib/src/util/log"
+import { notNullOrUndefined } from "@streamlit/lib/src/util/utils"
+
+import {
+  dataIsAnAppendOfPrev,
+  getDataArray,
+  getDataArrays,
+  getDataSets,
+  getInlineData,
+  VegaLiteChartElement,
+  WrappedNamedDataset,
+} from "./arrowUtils"
+import { useVegaLiteSelections } from "./useVegaLiteSelections"
+
+const DEFAULT_DATA_NAME = "source"
+
+interface UseVegaEmbedOutput {
+  createView: (
+    containerRef: RefObject<HTMLDivElement>,
+    spec: any
+  ) => Promise<VegaView | null>
+  updateView: (
+    data: Quiver | null,
+    datasets: WrappedNamedDataset[]
+  ) => Promise<VegaView | null>
+  finalizeView: () => void
+}
+
+export function useVegaEmbed(
```

Comment:
nit: maybe add a brief docstring comment for all custom hooks in this PR describing what they do.